core/arch/aarch64/build_ldstex.c

```diff
@@ -0,0 +1,297 @@
+ * them into a macro-instruction, OP_ldstex. This is a temporary solution for
+ * i#1698 and is likely to be fragile. Known problems:
+ *
+ * - We only handle single-entry single-exit contiguous code blocks. (Usually
+ *   they are written as inline assembler so they do fit this pattern.)
+ * - If the block uses all of X0-X5 and the stolen register then we cannot
+ *   mangle it (so it is better not to recognise it at all).
+ * - The contents of an OP_ldstex cannot be instrumented.
+ * - If execution remains in an OP_ldstex then signal delivery may be delayed.
+ * - Bad things might happen if there is a SIGSEGV or SIGBUS in an OP_ldstex.
+ * - Code flushing.
+ *
+ * This is currently modularised as a layer between the normal decoder and the
+ * block builder. It might be better to merge it with the block builder.
+ *
+ * If this solution can be made robust then it might be worth porting it to
+ * ARM/AArch32.
+ */
+
+#include "../globals.h"
+#include "arch.h"
+#include "decode.h"
+#include "disassemble.h"
+#include "instr.h"
+#include "instr_create.h"
+
+#include "build_ldstex.h"
+
+static bool
+instr_is_exclusive_load(instr_t *instr)
+{
+    switch (instr_get_opcode(instr)) {
+    case OP_ldaxp:
+    case OP_ldaxr:
+    case OP_ldaxrb:
+    case OP_ldaxrh:
+    case OP_ldxp:
+    case OP_ldxr:
+    case OP_ldxrb:
+    case OP_ldxrh:
+        return true;
+    }
+    return false;
+}
+
+static bool
+instr_is_nonbranch_pcrel(instr_t *instr)
+{
+    int i, n;
+    n = instr_num_dsts(instr);
+    for (i = 0; i < n; i++)
+        ASSERT(!OPND_IS_REL_ADDR(instr_get_dst(instr, i)));
+    n = instr_num_srcs(instr);
+    for (i = 0; i < n; i++)
+        if (OPND_IS_REL_ADDR(instr_get_src(instr, i)))
+            return true;
+    return false;
+}
+
+static void
+instr_create_ldstex(dcontext_t *dcontext, int len, uint *pc, instr_t *instr,
+                    OUT instr_t *instr_ldstex)
+{
+    int num_dsts = 0;
+    int num_srcs = 0;
+    int i, d, s, j;
+
+    for (i = 0; i < len; i++) {
+        ASSERT(instr[i].length == AARCH64_INSTR_SIZE &&
+               instr[i].bytes == instr[0].bytes + AARCH64_INSTR_SIZE * i);
+        num_dsts += instr_num_dsts(&instr[i]);
+        num_srcs += instr_num_srcs(&instr[i]);
+    }
+    instr_set_opcode(instr_ldstex, OP_ldstex);
+    instr_set_num_opnds(dcontext, instr_ldstex, num_dsts, num_srcs);
+    d = 0;
+    s = 0;
+    for (i = 0; i < len; i++) {
+        int dsts = instr_num_dsts(&instr[i]);
+        int srcs = instr_num_srcs(&instr[i]);
+        for (j = 0; j < dsts; j++)
+            instr_set_dst(instr_ldstex, d++, instr_get_dst(&instr[i], j));
+        for (j = 0; j < srcs; j++)
+            instr_set_src(instr_ldstex, s++, instr_get_src(&instr[i], j));
+    }
+    ASSERT(d == num_dsts && s == num_srcs);
+    /* Set pointer to original encoding. */
+    instr_ldstex->length = len * AARCH64_INSTR_SIZE;
+    instr_ldstex->bytes = instr[0].bytes;
```

Comment:
Use the interface: instr_set_raw_bits() will set both of these and other flags that the IR routines use.